### PR TITLE
Add multiple printers

### PR DESCRIPTION
**Pull request by josedonizetti**
### 1. Explain what the PR does

This PR add support for multiple printers to new `tracee` binary. The old binary `tracee-ebpf` maintain the previous syntax, and the change is only implemented for the new binary, for such I have first created a commit where I rename the previous `output` flag in order to simplify the deletion once we deprecate and remove `tracee-ebpf`. 

The new syntax was discussed on a previous PR https://github.com/aquasecurity/tracee/pull/2709
This PR is part of https://github.com/aquasecurity/tracee/issues/2355
Close https://github.com/aquasecurity/tracee/issues/2231

### 2. Explain how to test it

Please check the help commands:
```
tracee-ebpf --output help
tracee --output help
```

### 3. Other comments

Multiple printers will be the base of the webhook implementation for the new binary, where webhooks will be a printer like stdout, etc.

### CODE REVIEW

File: pkg/cmd/flags/output.go [hunk 1]
```diff
@@ -30,92 +34,82 @@ option:{stack-addresses,exec-env,relative-time,exec-hash,parse-arguments,sort-ev
...
   sort-events                                      enable sorting events before passing to them output. This will decrease the overall program efficiency.
+
 Examples:
   --output json                                            | output as json to stdout
-  --output gotemplate=/path/to/my.tmpl                     | output as the provided go template
-  --output out-file:/my/out --output log-file:/my/log      | output to /my/out and logs to /my/log
+  --output json:/my/out                                    | output as json to /my/out
+  --output gotemplate=/path/to/my.tmpl                     | output as the provided go template to stdout
+  --output gob:/my/out --output log-file:/my/log      	   | output gob to /my/out and logs to /my/log
+  --output json --output gob:/my/out                       | output as json to /my/out
```
Comment (geyslan) [inline pkg/cmd/flags/output.go:43]: json and gob?

File: pkg/cmd/flags/output.go [hunk 2]
```diff
@@ -30,92 +34,82 @@ option:{stack-addresses,exec-env,relative-time,exec-hash,parse-arguments,sort-ev
...
+  --output json:/my/out1,/my/out2                          | output as json to both /my/out and /my/out2
```
Comment (geyslan) [inline pkg/cmd/flags/output.go:44]: Tested as:

`--output json:./out1,./out2`
`--output json:stdout,./output`
`--output json:./output --output json`

And one of them doesn't outputs; it's outputting to one end only.

File: pkg/cmd/urfave/urfave.go [hunk 3]
```diff
@@ -101,9 +111,17 @@ func GetTraceeRunner(c *cli.Context, version string) (cmd.Runner, error) {
 
 	// Container information printer flag
 
-	printerConfig := output.PrinterConfig
-	printerConfig.ContainerMode = cmd.GetContainerMode(cfg)
-	cfg.Output = output.TraceeConfig
+	printers := make([]printer.EventPrinter, 0, len(output.PrinterConfigs))
+	for _, pConfig := range output.PrinterConfigs {
+		pConfig.ContainerMode = getContainerMode(cfg)
```
Comment (geyslan) [inline pkg/cmd/urfave/urfave.go:116]: I think you can use here the `(fs *FilterScopes) ContainerFilterEnabled()` helper.

**Review by geyslan — COMMENTED** [review]
I put some comments to be addressed.

File: pkg/cmd/flags/output.go (hunk 2 above)
Comment (josedonizetti) [reply pkg/cmd/flags/output.go:44]: @geyslan I've tested the samples above and they all worked. 🤔

File: pkg/cmd/urfave/urfave.go (hunk 3 above)
Comment (josedonizetti) [reply pkg/cmd/urfave/urfave.go:116]: @geyslan The `getContainerMode`is using `ContainerFilterEnabled()`helper, right?

File: pkg/cmd/flags/output.go [hunk 4]
```diff
@@ -13,16 +14,19 @@ import (
...
+Format options:
+table[:/path/to/file,...]                          output events in table format (default). The default path to file is stdout.
+table-verbose[:/path/to/file,...]                  output events in table format with extra fields per event. The default path to file is stdout.
+json[:/path/to/file,...]                           output events in json format. The default path to file is stdout.
+gob[:/path/to/file,...]                            output events in gob format. The default path to file is stdout.
+gotemplate=/path/to/template[:/path/to/file,...]   output events formatted using a given gotemplate file. The default path to file is stdout.
 none                                               ignore stream of events output, usually used with --capture
+
+Log options:
+log-file:/path/to/file                             write the logs to a specified file. create/trim the file if exists (default: stderr)
```
Comment (yanivagman) [inline pkg/cmd/flags/output.go:26]: log options should probably move to the `--log` flag

File: pkg/cmd/flags/output.go [hunk 5]
```diff
@@ -13,16 +14,19 @@ import (
...
+forward:url                                        send events in json format using the Forward protocol to a Fluent receiver
```
Comment (yanivagman) [inline pkg/cmd/flags/output.go:27]: is this a log option (since it is located here) or an output printer?

Comment (josedonizetti) [reply pkg/cmd/flags/output.go:27]: @yanivagman output printer.

File: pkg/cmd/flags/output.go (hunk 4 above)
Comment (josedonizetti) [reply pkg/cmd/flags/output.go:26]: I like it, will create an issue for it.

File: pkg/cmd/urfave/urfave.go (hunk 3 above)
Comment (geyslan) [reply pkg/cmd/urfave/urfave.go:116]: Actually you're using the FilterScope helper, I've mentioned the FilterScope**S** one, which already provides the needed information based on a bitmask filled through (fs *FilterScopes) Compute().

File: pkg/cmd/printer/broadcast.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+func (b *Broadcast) Print(event trace.Event) {
+	for _, c := range b.eventsChan {
+		// we use select to avoid blocking the print loop if one of the printers is slow
+		// in case the event can't be sent because the buffered channel is full, we drop it
+		// and log an error
+		select {
+		case c <- event:
+		default:
+			logger.Error("dropping event due to slow printer", "event", event)
+		}
```
Comment (geyslan) [inline pkg/cmd/printer/broadcast.go:40]: I got more error logs than events, running:

sudo ./dist/tracee -t e='*' --output table:stdout

This logger Error polutes table output, WDYT of changing it to debug level and to increment errors channel?

**Review by geyslan — CHANGES_REQUESTED** [review]

File: pkg/cmd/printer/broadcast.go (hunk 6 above)
Comment (josedonizetti) [reply pkg/cmd/printer/broadcast.go:40]: @geyslan Done. Though if you test with "e='*'" we will lose events just like we do on the main branch, and possibly the  files generated with two prints will have different event on it.

**Review by geyslan — APPROVED** [review]
LGTM. Great job with the printer broadcast.

File: pkg/cmd/printer/broadcast.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+func (b *Broadcast) Print(event trace.Event) {
+	for _, c := range b.eventsChan {
+		// we use select to avoid blocking the print loop if one of the printers is slow
+		// in case the event can't be sent because the buffered channel is full, we drop it
+		// and log an warning
+		select {
+		case c <- event:
+		default:
+			// TODO: add metrics about not printed events
+			logger.Debug("dropping event due to slow printer", "event", event)
```
Comment (yanivagman) [inline pkg/cmd/printer/broadcast.go:41]: I think this deserves at least info level log message

File: pkg/cmd/urfave/urfave.go (hunk 3 above)
Comment (yanivagman) [inline pkg/cmd/urfave/urfave.go:116]: `getContainerMode(cfg)` can be called once outside of the loop

**Review by yanivagman — APPROVED** [review]
Nice work @josedonizetti !
Overall LGTM, some nit comments ahead

File: pkg/cmd/printer/broadcast.go (hunk 7 above)
Comment (geyslan) [reply pkg/cmd/printer/broadcast.go:41]: Info is default level, so depending on the volume of events it will output more logs than events.

Comment (yanivagman) [reply pkg/cmd/printer/broadcast.go:41]: I think this should happen too frequently (at least I hope so), and if something is dropped here, there should be a way for the user to be aware of that. So as long as we don't have metrics for events dropped here, we should expose this information somehow

**Comment by rafaeldtinoco** [comment]
Make sure to rebase so https://github.com/aquasecurity/tracee/pull/2776 is included and new testers are used.

**Comment by rafaeldtinoco** [comment]
Thanks for waiting on my review/read. Nice job!

**Review by rafaeldtinoco — COMMENTED** [review]
LGTM

**Pull request merged**